### ISSUE TITLE
Error merging files during optimization

### ISSUE THREAD

**Issue by plaisted** [pdfcpu/pdfcpu#98]
Encountered an error testing the tool for merging some PDFs I have.  The PDF did validate using your tool (and others).

```
 INFO: 2019/08/01 22:55:57 optimizing fonts & images
Fatal: equalObjects: unhandled compare for type <nil>

github.com/hhrutter/pdfcpu/pkg/pdfcpu.equalObjects
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/equal.go:67
github.com/hhrutter/pdfcpu/pkg/pdfcpu.equalArrays
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/equal.go:81
github.com/hhrutter/pdfcpu/pkg/pdfcpu.equalObjects
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/equal.go:64
github.com/hhrutter/pdfcpu/pkg/pdfcpu.equalDicts
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/equal.go:181
github.com/hhrutter/pdfcpu/pkg/pdfcpu.equalStreamDicts
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/equal.go:96
github.com/hhrutter/pdfcpu/pkg/pdfcpu.handleDuplicateImageObject
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/optimize.go:335
github.com/hhrutter/pdfcpu/pkg/pdfcpu.optimizeXObjectResourcesDict
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/optimize.go:408
github.com/hhrutter/pdfcpu/pkg/pdfcpu.optimizeResources
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/optimize.go:493
github.com/hhrutter/pdfcpu/pkg/pdfcpu.parseResourcesDict
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/optimize.go:520
github.com/hhrutter/pdfcpu/pkg/pdfcpu.parsePagesDict
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/optimize.go:587
github.com/hhrutter/pdfcpu/pkg/pdfcpu.parsePagesDict
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/optimize.go:568
github.com/hhrutter/pdfcpu/pkg/pdfcpu.parsePagesDict
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/optimize.go:568
github.com/hhrutter/pdfcpu/pkg/pdfcpu.parsePagesDict
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/optimize.go:568
github.com/hhrutter/pdfcpu/pkg/pdfcpu.parsePagesDict
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/optimize.go:568
github.com/hhrutter/pdfcpu/pkg/pdfcpu.optimizeFontAndImages
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/optimize.go:730
github.com/hhrutter/pdfcpu/pkg/pdfcpu.OptimizeXRefTable
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/optimize.go:1054
github.com/hhrutter/pdfcpu/pkg/api.OptimizeContext
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/api/api.go:106
github.com/hhrutter/pdfcpu/pkg/api.Merge
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/api/api.go:1238
github.com/hhrutter/pdfcpu/pkg/api.MergeFile
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/api/api.go:1285
github.com/hhrutter/pdfcpu/pkg/cli.Merge
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/cli/cli.go:121
github.com/hhrutter/pdfcpu/pkg/cli.Process
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/pkg/cli/process.go:90
main.process
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/cmd/pdfcpu/main.go:225
main.handleMergeCommand
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/cmd/pdfcpu/prepare.go:163
main.CommandMap.Handle
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/cmd/pdfcpu/registry.go:85
main.main
        /Users/horstrutter/go/src/github.com/hhrutter/pdfcpu/cmd/pdfcpu/main.go:181
runtime.main
        /Users/horstrutter/gotip/src/runtime/proc.go:200
runtime.goexit
        /Users/horstrutter/gotip/src/runtime/asm_amd64.s:1337
```

It's a very large confidential PDF I was doing some performance testing on so can't share it but believe based on the stacktrace it's coming from this type of stream dictionary in the PDF:
```
<</Type/XObject/Subtype/Image/Width 150/Height 128/BitsPerComponent 1/ColorSpace/DeviceGray/Decode[0 1]/Filter[/ASCII85Decode/CCITTFaxDecode]/DecodeParms[null<</K -1/Columns 150/Rows 128>>]/Length 786>>stream
```
`equalObjects` does not have a case for Null objects (error shows `<nil>` type, is that how pdfcpu internally represents pdf null objects?).

PDF 1.7 spec specifically states to use the null object in this scenario:
```
DecodeParms must be an array with one entry for each filter: either the parameter dictionary for that filter, or the null object if that filter has no parameters (or if all of its parameters have their default values).
```

Thanks for putting the work into the tool! Not currently using it but try to keep my eye on all the good PDF libraries in case the need arises.

**Comment by hhrutter** [comment]
Thanks for reporting this.
Filter pipelines are not used that often but that sure smells like a bug.

**Comment by hhrutter** [comment]
Is there any way you can provide a sample?

**Comment by plaisted** [comment]
I'll see if I can get a reproducible sample without any confidential data.

**Comment by hhrutter** [comment]
`equalObjects` was fixed for handling `NULL` with [4d4296a](https://github.com/pdfcpu/pdfcpu/commit/4d4296aa55d0f0600cf5b5a8a5947d9dadc03f31).
